File: DLGPcpp/dlgp.cpp
#include "dlgp.h"
using Eigen::seq;

dlgp::dlgp(int xSize0, int outs0 ,int pts0, int N0, bool ard0){
	// initialize data
	//see parameter description in dlgp.h
	pts = pts0;
	N = N0;
	xSize = xSize0;
	outs = outs0;
	count = 0;

	X.resize(xSize, pts * N);	X.setZero();
	Y.resize(outs, pts * N);	Y.setZero();
	K.resize(pts * outs, pts * N);	K.setZero();
	//invK.resize(pts * outs, pts * N);	invK.setZero();
	alpha.resize(pts * outs, N);	alpha.setZero();
	
	localCount.resize( 2 * N - 1);	localCount.setZero();
	medians.resize( 2 * N - 1);
	auxUbic.resize( 2 * N - 1);
	auxUbic(0) = 0;
	auxUbic(2 * N - 2) = -1;

	children.resize(2, 2 * N - 1);	children.setConstant(-1);
	
	overlaps.resize(2, 2 * N - 1);

	if (ard0) {
		amountL = xSize;
	}
	else
	{
		amountL = 1; 
	}
}

Eigen::RowVectorXd dlgp::kernel(Eigen::MatrixXd Xi, Eigen::VectorXd Xj, int out) {
	// Compute the squared exponential kernel (aka RBF)
	Eigen::RowVectorXd kern(Xi.cols());
	for (int i = 0; i < Xi.cols(); i += 1) {
		kern(i) = std::pow(sigmaF(out), 2) * exp( -0.5 * ((Xi.col(i)-Xj).array()/lengthS( seq((out)*amountL,(out+1)*amountL-1)) ).pow(2).sum());
	}
	return kern;
}

double mValue(int model, int cutD);

void dlgp::updateParam(Eigen::VectorXd x, int model) {
	localCount(model) += 1;
	int pos = auxUbic(model);
	if (localCount(model) == 1) {	//set parameters for the first datum
		for (int p = 0; p < outs; p += 1) {
			double kVal = kernel(x, x, p)(0);
			K(p * pts, pos * pts) = kVal + std::pow(sigmaN(p), 2);
			alpha(p * pts, pos) = Y(p, pos * pts) / kVal;
		}
	}
	else {
		//update parameters for more data
		Eigen::MatrixXd auxX = X( Eigen::all, seq (pos*pts, pos*pts+localCount(model) - 2) ); // does not consider newest pt
		Eigen::MatrixXd auxY = Y( Eigen::all, seq(pos * pts, pos * pts + localCount(model) - 1) );
		for (int p = 0; p < outs; p += 1) {
			Eigen::MatrixXd b = kernel(auxX, x, p);
			double c = kernel(x, x, p)(0) + std::pow(sigmaN(p),2);
			int auxOut = p * pts;

			
			K( auxOut + localCount(model) - 1 , seq(pos * pts, pos * pts + localCount(model) - 2 ) ) = b;
			K( seq(auxOut, auxOut + localCount(model) - 2) , pos * pts + localCount(model) - 1 ) = b.transpose();
			K( auxOut + localCount(model) - 1 , pos * pts + localCount(model) - 1 ) = c;
			
			
			alpha( seq(auxOut,auxOut+localCount(model) - 1) , pos ) = K( seq(auxOut,auxOut+localCount(model)-1) , 
				seq(pos*pts,pos*pts+localCount(model)-1)  ).ldlt().solve( auxY( p , Eigen::all ).transpose() );
			
		}
	}
}

void dlgp::addPoint(Eigen::VectorXd x, Eigen::VectorXd y, int model) {
	if (localCount(model) < pts) { // if model is not full
		X(Eigen::all, auxUbic(model) * pts + localCount(model) ) = x;
		Y(Eigen::all, auxUbic(model) * pts + localCount(model) ) = y;
		updateParam(x, model);
	}
	if (localCount(model) == pts) {
		divide(model);
	}
}

void dlgp::divide(int model) {
	if (auxUbic(Eigen::last) != -1) {//(parent(Eigen::last) != 0 )
		std::cout << "no room for more divisions" << std::endl << std::endl;
		return;
	}
	//obtain cutting dimension
	Eigen::MatrixXd::Index cutD;
	double width = ( X(Eigen::all, seq(auxUbic(model) * pts, auxUbic(model) * pts + pts - 1)).rowwise().maxCoeff() - 
		X(Eigen::all, seq(auxUbic(model) * pts, auxUbic(model) * pts + pts - 1)).rowwise().minCoeff() ).maxCoeff(&cutD);
	//obtain hyperplane with (max + min) /2
	double mP = ( X(cutD, seq(auxUbic(model) * pts, auxUbic(model) * pts + pts - 1)).maxCoeff() +
		X(cutD, seq(auxUbic(model) * pts, auxUbic(model) * pts + pts - 1)).minCoeff() ) / 2;
	//set overlapping region and median
	double o = width / wo; 
	if (o == 0)
		o = 0.1;

	medians(model) = mP;
	overlaps(0, model) = cutD;
	overlaps(1, model) = o;
	//matrices for left/right models
	Eigen::MatrixXd xL(xSize,pts) , xR(xSize, pts);
	Eigen::MatrixXd yL(outs, pts) , yR(outs, pts);

	int lcount = 0;
	int rcount = 0;
	Eigen::VectorXi iL(pts), iR(pts); //vector with indeces

	for (int i = 0; i < pts; i += 1) { //sort the data
		double xD = X(cutD,  auxUbic(model) * pts + i); 
		if (xD < mP - 0.5 * o){ // if in left
			xL(Eigen::all, lcount) = X(Eigen::all, auxUbic(model) * pts + i);
			yL(Eigen::all, lcount) = Y(Eigen::all, auxUbic(model) * pts + i);
			iL(lcount) = i;
			lcount += 1;
		}
		else if (xD >= mP - 0.5*o && xD <= mP + 0.5*o ){ // if in overlapping
			double pL = 0.5 + (xD - mP) / o;
			if (pL >= (double)((rand() % 10000) + 1) / 10000) { //select left
				xL(Eigen::all, lcount) = X(Eigen::all, auxUbic(model) * pts + i);
				yL(Eigen::all, lcount) = Y(Eigen::all, auxUbic(model) * pts + i);
				iL(lcount) = i;
				lcount += 1;
			}
			else { //select right
				xR(Eigen::all, rcount) = X(Eigen::all, auxUbic(model) * pts + i);
				yR(Eigen::all, rcount) = Y(Eigen::all, auxUbic(model) * pts + i);
				iR(rcount) = i;
				rcount += 1;
			}
		}
		else if (xD > mP + 0.5*o) { //if in right
			xR(Eigen::all, rcount) = X(Eigen::all, auxUbic(model) * pts + i);
			yR(Eigen::all, rcount) = Y(Eigen::all, auxUbic(model) * pts + i);
			iR(rcount) = i;
			rcount += 1;
		}
	}
	localCount(model) = 0; //divided model is now empty
	if (count == 0)
		count += 1;
	else
		count += 2;
	children(0, model) = count; children(1, model) = count + 1; //assign children
	//set parameters of new models
	localCount(count) = lcount; localCount(count + 1) = rcount;
	auxUbic(count) = auxUbic(model); auxUbic(count + 1) = auxUbic.maxCoeff()+1;

	//values for K permutation
	Eigen::VectorXi order(pts);
	order << iL(seq(0, lcount - 1)), iR(seq(0, rcount - 1));
	Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic> perm(pts);
	perm.indices() = order.array();
	
	for (int p = 0; p < outs; p += 1) { //compute parameters of new model
		Eigen::MatrixXd newK = K(seq(p * pts, (p + 1) * pts -1), seq(auxUbic(model) * pts, auxUbic(model) * pts + pts -1));
		//permute K
		newK = perm.transpose() * newK *perm;
		//std::cout << newK << std::endl;
		//set child K
		K(seq(p * pts, p * pts + lcount - 1), seq(auxUbic(count) * pts, auxUbic(count) * pts + lcount - 1)) =
			newK(seq(0, lcount-1), seq(0, lcount-1));
		K(seq(p * pts, p * pts + rcount - 1), seq(auxUbic(count+1) * pts, auxUbic(count+1) * pts + rcount - 1)) =
			newK(seq(lcount, Eigen::last), seq(lcount, Eigen::last));
		//set child alphas
		alpha(seq(p * pts, p * pts + lcount - 1), auxUbic(count)) =
			newK(seq(0, lcount-1), seq(0, lcount-1)).ldlt().solve(yL(p, seq(0, lcount - 1)).transpose());
		alpha(seq(p * pts, p * pts + rcount - 1), auxUbic(count+1)) =
			newK(seq(lcount, Eigen::last), seq(lcount, Eigen::last)).ldlt().solve(yR(p, seq(0, rcount - 1)).transpose());
	}

	auxUbic(model) = -1; // parent set will not have more data
	//relocate X and Y
	X(Eigen::all, seq(auxUbic(count) * pts, auxUbic(count) * pts + pts - 1)) = xL;
	X(Eigen::all, seq(auxUbic(count+1) * pts, auxUbic(count+1) * pts + pts - 1)) = xR;
	Y(Eigen::all, seq(auxUbic(count) * pts, auxUbic(count) * pts + pts - 1)) = yL;
	Y(Eigen::all, seq(auxUbic(count + 1) * pts, auxUbic(count + 1) * pts + pts - 1)) = yR;
}

double dlgp::activation(Eigen::VectorXd x, int model) {
	if (children(1, model) == -1) //return zero when model is a leaf
		return 0;
	double mP = medians(model); //hyperplane value
	double xD = x((int)overlaps(0, model)); //value of x in cutting dimension
	double o = overlaps(1, model); //overlapping region
	if (xD < mP - 0.5 * o)
		return 1;
	else if (xD >= mP - 0.5 * o && xD <= mP + 0.5 * o) // if in overlapping
		return 0.5 + (xD - mP) / o;
	else
		return 0;
}

void dlgp::update(Eigen::VectorXd x, Eigen::VectorXd y) {
	int model = 0;
	if (localCount(3) == 48)
		int a = 0;
	while (children(1, model) != -1) { // if model is a parent
		// search for a leaf to asign the point
		double pL = activation(x, model);
		if (pL >= (double)((rand() % 10000) + 1) / 10000)
			model = children(0, model);// left child
		else
			model = children(1, model);
	}
		// add the model to the randomly selected model
	addPoint(x, y, model);
}

Eigen::VectorXd dlgp::predict(Eigen::VectorXd x) {
	Eigen::VectorXd out(outs);	out.setConstant(0);
	Eigen::VectorXi models(1000); // active models
	Eigen::VectorXd probs(1000); // global probs. 

	models(0) = 0; //start in root
	models(1) = 0;
	probs(0) = 1;
	probs(1) = 1;
	int mCount = 1; //amount of GPs used for prediction
	//while (children(0, models(mCount-1)) != -1 && children(0, models( mCount)) != -1) {
	while (children(0, models(seq(0,mCount - 1))).array().sum() != -1 * mCount)  {
		for (int j = 0; j < mCount; j += 1) {
			if (children(0, models(j)) != -1) { // go deeper in tree if node has children
				double pL = activation(x, models(j));
				if (pL == 1)
					models(j) = children(0, models(j));
				else if (pL == 0)
					models(j) = children(1, models(j));
				else if (1 > pL > 0) { //both children have non-zero probs.
					mCount += 1;

					models(mCount - 1) = children(1, models(j));
					probs(mCount - 1) = probs(1, j) * (1 - pL);

					models(j) = children(0, models(j));
					probs(j) = probs(j) * pL;
				}
			}
		}
	}
	//prediction weighting predictions
	Eigen::VectorXd pred(1);
	for (int p = 0; p < outs; p += 1) {
		for (int i = 0; i < mCount; i += 1) {
			int model = models(i);
			pred = kernel(X(Eigen::all, seq(auxUbic(model) * pts, auxUbic(model) * pts + localCount(model) - 1)), x, p)*
				alpha(seq(p*pts, p*pts+localCount(model)-1), auxUbic(model) );
			out(p) += pred(0) * probs(i);
		}
	}
	return out;

}

File: DLGPpy/dlgp.py
import numpy as np
import random


class dlgp:
    def __init__(self, xSize, outs, pts, N, ard):
        #user defined values
        self.pts = pts #data limit per local GP
        self.N = N #max. number of leaves
        self.xSize = xSize #dimensionality of X
        self.outs = outs #dimensionality of Y
        #parameters
        self.wo = 300 #factor width / overlapping
        
        #auto. hyp. optimization
        self.sigmaF = np.ones([outs, N] , dtype = float)
        self.sigmaN = np.ones([outs, N] , dtype = float)
        
        if ard:
            self.lengthS = np.ones([xSize * outs , N ] ,dtype = float)
            self.amountL = xSize
            #hyperparameter optimization variables
            self.dlik0 = 0.1 * np.ones([(xSize + 2) * outs, N], dtype = float) #local gradients
            self.delta = np.ones([(xSize + 2) * outs, N], dtype = float) #local stepsize
        else:
            self.lengthS = 1
            self.amountL = 1
        #data
        self.X = np.zeros( [xSize , pts * N] , dtype = float )
        self.Y = np.zeros( [outs , pts * N] , dtype = float )
        self.K = np.zeros( [pts*outs , pts * N] , dtype = float ) #covariance matrix
        self.alpha = np.zeros( [pts*outs , N] , dtype = float ) #prediction vectors
        #aux variables
        self.count = 0 #number of leaves - 1
        self.localCount = np.zeros( 2 * N - 1 , dtype = int ) #pts in each set
        self.medians = np.zeros( 2 * N - 1 , dtype = float ) #vector of hyperplanes
        self.overlapD = np.zeros( 2 * N - 1 , dtype = int) #overlap dimension
        self.overlapW = np.zeros( 2 * N - 1 , dtype = float) #overlap width
        self.auxUbic = np.zeros( 2 * N - 1 , dtype = int ) - 1 #map the position of data
        self.auxUbic[0] = 0
        self.children = np.zeros( [2, 2 * N - 1] , dtype = int) -1 #line 1: left child, line 2: right child

        
    def kernel(self, Xi, Xj, out): #returns a scalar or a vector
        if Xi.ndim == 1:
            kern =  (self.sigmaF[out]**2) * np.exp( -0.5* np.sum(((Xi - Xj)/self.lengthS[:,out])**2)) 
            return kern
        else:
            kern = (self.sigmaF[out]**2) * np.exp(-0.5*np.sum(((Xi.transpose() - Xj)/self.lengthS[:,out])**2, axis = 1))
            return kern
        
    def fkernel(Xi, hyp): #returns a kernel matrix of Xi and itself
    #hyp is [sigmaN, sigmaN, sigmaL_1, sigmaL_2, ... , sigmaL_xSize]
        kern = np.zeros( [ Xi.shape[1] , Xi.shape[1] ], dtype = float)
        if hyp.shape[0] > 3:
            for p in range (hyp.shape[0] - 2):
                k1, _ = np.meshgrid(Xi[p, :], Xi[p, :])
                print(p)
                kern = (kern + ( k1 - np.transpose(k1) )**2 ) / (hyp[p+1]**2)
        else:
            for p in range(Xi.shape[0]):
                k1, _ = np.meshgrid(Xi[p, :], Xi[p, :])
                print(k1)
                kern = (kern + ( k1 - np.transpose(k1) )**2 ) / (hyp[2]**2)
        kern = (hyp[0]**2) * np.e ** (-0.5 * kern)
        return kern
    
    def rprop(self, model):
        for p in range(self.outs):
            #get previous hyperparameters
            x0 = np.append( self.sigmaF[p+1 , self.auxUbic[model] ] ,   \
                           self.sigmaF[p+1 , self.auxUbic[model] ] , \
                               self.lengthS[ p * self.amountL + 1 : (p+1) * self.amountL , \
                                            self.auxUbic[model]])
            dmax = 50
            dmin = 1e-6
            etap = 1.2
            etam = 0.5
            #get new hyperparameters
            x1 = x0 + np.sign( self.dlik0[ p*(self.amountL+2) + 1 : \
                        (p+1)*(self.amountL+2) , self.auxUbic[model]  ] * \
                              self.delta[ p*(self.amountL+2)+1:\
                                    (p+1)*(self.amountL+2),self.auxUbic[model]])
            #sigmaN and sigmaF value restricions
            if abs(x1[1])  < 0.01:
                x1[1] = np.sign(x1[1])*.01
            if np.abs(x1[0] / x1[1]) > 1000:
                x1[0] = np.sign(x1[0]) * x1[1] * 1000
            #get the gradient    
            dfx1 = self.grad(x1, model, p)
            #update delta and check if it fulfills requirements
            dfx01 = self.dlik0[p*(self.amountL+2)+1 : (p+1)*(self.amountL+2),\
                self.auxUbic[model]]*dfx1 #product of gradients
            delta0 = self.delta[ p*(self.amountL+2)+1 : (p+1)*(self.amountL+2) \
                ,self.auxUbic[model]] #previous delta
            d1 = (dfx01>0)*delta0*etap + (dfx01<0)*delta0*etam + (dfx01==0)*delta0*etap
            d1 = d1*(d1>=dmin & d1<=dmax) + (d1<dmin)*dmin + (d1>dmax)*dmax
            
            #update properties
            self.dlik0[p*(self.amountL+2)+1 : (p+1)*(self.amountL+2),\
                self.auxUbic[model]] = dfx1
            self.delta[ p*(self.amountL+2)+1 : (p+1)*(self.amountL+2) \
                ,self.auxUbic[model]] = d1
            self.sigmaF[p+1, self.auxUbic[model]] = x1[0]
            self.sigmaN[p+1, self.auxUbic[model]] = x1[1]
            self.lengthS[p*self.amountL+1:(p+1)*self.amountL, self.auxUbic[model]]=\
                x1[2:]
                
    def updateParam(self, x, model):
        self.localCount[model] += 1
        pos = self.auxUbic[model]
        if self.localCount[model] == 1:
            for p in range(self.outs):
                kVal = self.kernel(x,x,p)
                self.K[ p*self.pts , pos*self.pts ] = kVal + self.sigmaN[p]**2
                self.alpha[ p * self.pts , pos ] =  self.Y[ p,  pos * self.pts ] /kVal
        else:
            auxX = self.X[ : , pos*self.pts : pos * self.pts + self.localCount[model]  ] #does not include x
            auxY = self.Y[ : , pos*self.pts : pos * self.pts + self.localCount[model]  ]
            for p in range(self.outs):
                b = self.kernel( auxX , x, p )
                b[-1] += self.sigmaN[p] ** 2
                auxOut = p * self.pts
                
                self.K[auxOut + self.localCount[model] - 1 ,
                       pos * self.pts : pos * self.pts + self.localCount[model] -1] = b[0:-1]
                self.K[ auxOut : auxOut + self.localCount[model] ,
                       pos * self.pts + self.localCount[model] -1 ] = np.transpose(b)
                
                self.alpha[ auxOut : auxOut + self.localCount[model] ,  pos ] = \
                    np.linalg.solve( self.K[ auxOut : auxOut + self.localCount[model] ,
                           pos*self.pts : pos * self.pts + self.localCount[model] ] , auxY[ p , : ] )

    def addPoint(self, x, y,model):
        if self.localCount[model] < self.pts:
            self.X[ : , self.auxUbic[model] * self.pts + self.localCount[model] ] = x 
            self.Y[ : , self.auxUbic[model] * self.pts + self.localCount[model] ] = y 
            self.updateParam(x , model)
        if self.localCount[model] == self.pts:
            self.divide(model)
    
    def divide(self, model):
        if self.auxUbic[-1] != -1:
            print( "no room for more divisions" )
            return
        #compute widths in all dimensions
        width = self.X[ : , self.auxUbic[ model ] * self.pts : self.auxUbic[ model ] *
            self.pts + self.pts ].max(axis = 1) - self.X[ : , self.auxUbic[ model ] *
                self.pts : self.auxUbic[ model ] * self.pts + self.pts ].min(axis = 1)
                                                             
        #obtain cutting dimension
        cutD = np.argmax(width)
        width = width.max()
        #compute hyperplane
        mP = (self.X[ cutD , self.auxUbic[ model ] * self.pts : self.auxUbic[ model ] * 
            self.pts + self.pts].max() + self.X[ cutD , self.auxUbic[ model ] *
                self.pts : self.auxUbic[ model ] * self.pts + self.pts ].min())/2
        
        #get overlapping region
        o = width / self.wo
        if o == 0:
            o = 0.1
        
        self.medians[model] = mP  #set model hyperplane
        self.overlapD[model] = cutD  #cut dimension
        self.overlapW[model] = o  #width of overlap
        
        xL = np.zeros( [self.xSize, self.pts] , dtype = float )
        xR = np.zeros( [self.xSize, self.pts] , dtype = float )
        yL = np.zeros( [self.outs, self.pts] , dtype = float )
        yR = np.zeros( [self.outs, self.pts] , dtype = float )
        
        lcount = 0
        rcount = 0
        
        iL = np.zeros( self.pts, dtype = int )
        iR = np.zeros( self.pts, dtype = int )
        
        for i in range(self.pts):
            xD = self.X[ cutD , self.auxUbic[model] * self.pts + i ]
            if xD < mP - 0.5 * o:
                xL[:, lcount] = self.X[:, self.auxUbic[model] * self.pts + i ]
                yL[:, lcount] = self.Y[:, self.auxUbic[model] * self.pts + i ]
                iL[lcount] = i
                lcount += 1
            elif xD >= mP - 0.5*o and xD <= mP + 0.5*o: #if in overlapping
                pL = 0.5 + (xD - mP) / o #prob. of being in left
                if pL >= random.random() and pL != 0: #left selected
                    xL[:, lcount] = self.X[:, self.auxUbic[model] * self.pts + i ]
                    yL[:, lcount] = self.Y[:, self.auxUbic[model] * self.pts + i ]
                    iL[lcount] = i
                    lcount += 1
                else:
                    xR[:, rcount] = self.X[:, self.auxUbic[model] * self.pts + i ]
                    yR[:, rcount] = self.Y[:, self.auxUbic[model] * self.pts + i ]
                    iR[rcount] = i
                    rcount += 1
            elif xD > mP + 0.5*o: #if in right
                xR[:, rcount] = self.X[:, self.auxUbic[model] * self.pts + i ]
                yR[:, rcount] = self.Y[:, self.auxUbic[model] * self.pts + i ]
                iR[rcount] = i
                rcount += 1
        self.localCount[model] = 0
        #update counter
        if self.count == 0:
            self.count+=1
        else:
            self.count+=2
        #assign children
        self.children[0,model] = self.count
        self.children[1,model] = self.count +1
        #set local count of children
        self.localCount[self.count] = lcount
        self.localCount[self.count + 1] = rcount 
        self.auxUbic[self.count] = self.auxUbic[model]
        self.auxUbic[self.count+1] = self.auxUbic.max() + 1
        #values for K permutation
        order = np.concatenate( (iL[0:lcount] , iR[0:rcount]) )
        #update parameters of child models
        for p in range(self.outs):
            newK = self.K[ p * self.pts : (p+1) * self.pts , self.auxUbic[model] * self.pts:
                self.auxUbic[model] * self.pts + self.pts]
            #permute K
            newK[ range(self.pts), : ] = newK[ order , : ]
            newK[ : , range(self.pts) ] = newK[ : , order ]
            #set child K
            self.K[p*self.pts : p * self.pts + lcount , self.auxUbic[self.count] * self.pts : 
                 self.auxUbic[self.count] * self.pts + lcount ] = newK[ 0 : lcount , 0 : lcount ]
            self.K[p*self.pts : p * self.pts + rcount  , self.auxUbic[self.count+1] * self.pts : 
                 self.auxUbic[self.count+1] * self.pts + rcount  ] = \
                    newK[lcount : self.pts,lcount : self.pts]
            #set child alpha
            self.alpha[ p * self.pts : p * self.pts + lcount , self.auxUbic[self.count]] = \
                np.linalg.solve(newK[ 0 : lcount , 0 : lcount ] , yL[p, 0:lcount].transpose())
            self.alpha[ p * self.pts : p * self.pts + rcount , self.auxUbic[self.count+1]] = \
                np.linalg.solve(newK[lcount : self.pts,lcount : self.pts], yR[p, 0:rcount].transpose())
        #parent will not have more data:
        self.auxUbic[model] = -1
        #relocate X Y to children
        self.X[: , self.auxUbic[self.count] * self.pts : 
               self.auxUbic[self.count] *self.pts + self.pts ] = xL
        self.X[: , self.auxUbic[self.count+1] * self.pts : 
               self.auxUbic[self.count+1] *self.pts + self.pts ] = xR  
        self.Y[: , self.auxUbic[self.count] * self.pts : 
               self.auxUbic[self.count] *self.pts + self.pts ] = yL
        self.Y[: , self.auxUbic[self.count+1] * self.pts : 
               self.auxUbic[self.count+1] *self.pts + self.pts ] = yR     
    
    def activation(self, x, model):
        if ( self.children[ 0 , model ] == -1 ): #return 0 if model is a leaf
            return 0
        mP = self.medians[model] #hyperplane value
        xD = x[ self.overlapD[model] ] #x value in cut dimension
        o = self.overlapW[model]    #overlap width
        if xD < mP - 0.5*o:
            return 1
        elif xD >= mP - 0.5*o  and xD <= mP + 0.5*o:
            return 0.5 + (xD - mP) / o
        else:
            return 0
        
    def update( self , x , y ):
        model = 0
        while self.children[0,model] != -1: #while model is parent
            #search for a leaf
            pL = self.activation( x , model )
            if pL >= random.random() and pL != 0:
                model = self.children[0,model] #go left
            else:
                model = self.children[1,model] #go right
        self.addPoint(x, y, model)
                
    def predict(self, x):
        out = np.zeros( self.outs , dtype = float )
        models = np.zeros( 1000 , dtype = int )
        probs = np.zeros( 1000 , dtype = float ) + 1
        
        mCount = 1
        while self.children[ 0 , models[0:mCount] ].sum() != -1 * mCount:
            for j in range( mCount ):
                if self.children[ 0 , models[j] ] != -1: #go deeper in three if node has children
                    pL = self.activation( x , models[j] )
                    if pL == 1:
                        models[j] = self.children[ 0 , models[j] ]
                    elif pL == 0:
                        models[j] = self.children[ 1 , models[j] ]
                    elif 1 > pL > 0:
                        mCount += 1
                        
                        models[ mCount - 1 ] = self.children[ 1 , models[j] ]
                        probs[ mCount - 1 ] = probs[j] * ( 1 - pL )
                        
                        models[j] = self.children[ 0 , models[j] ]
                        probs[j] = probs[j] * pL
        for p in range(self.outs):
            for i in range( mCount ):
                model = models[i]
                pred = np.dot(self.kernel( self.X[ : , self.auxUbic[model]*self.pts : \
                    self.auxUbic[model]*self.pts + self.localCount[model] ] , x , p ), 
                        self.alpha[ p*self.pts : p*self.pts + self.localCount[model] , 
                            self.auxUbic[model] ])
                out[p] += pred * probs[i]
        return out